# Subway Data Analysis

## Introduction

The NYC public transportantion system - Metro Transit Authority - provides data for download via csv files. Part of the information available are data from the subway turnstiles, containing weekly logs for cumulative entries and exits by turnstile and by subway station during a provided timeframe.


For this project, we will only use the information available at: http://web.mta.info/developers/turnstile.html.




# About this project

For this project, you will apply the knowledge acquired in the first month of this course. We will practice basic data acquisition and data cleaning tasks to find out fundamental stuff about the data using what we learned in the Statistics course. 

The goal of this project is to explore the relationship between data from the NYC Subway turnstiles and the city weather. For this, besides data from the subway, we will also need data from the weather in NYC. 

Here are the main points that will be considered in this work:

- Gathering data from the Internet
- Using Statistics for Data Analysis
- Data handling and simple graphics creation with `Pandas`

*How to find help*: We suggest that you try the following channels, in the following order:

| Type of Question\Channels    	| Google 	| Forum 	| Slack 	| Email 	|
|-------------------------------	|--------	|-------	|-------	|-------	|
| Pandas and Python Programming 	| 1      	| 2     	| 3     	|       	|
| Projects Requiriments         	|        	| 1     	| 2     	| 3     	|
| Projects Specific Parts       	|        	| 1     	| 2     	| 3     	|

Here is the address for each of these channels:

- Forum: https://discussions.udacity.com/c/ndfdsi-project
- Slack: [Big Data Foundations](https://goo.gl/4K7LWK)
- Email: india@udacity.com

**The student is expected to submit this report including:**

- All TODO's completed, as they are crucial for the code to run accordingly
- The ipynb file, exported as html

To submit this project, go to the [classroom](https://coco.udacity.com/nanodegrees/nd100-inbig/locale/en-us/versions/1.0.0/parts/469348/modules/469702/lessons/469703/project), and submit your zipped `.ipynb` and html.

# Reminders

Before we start, there are a few things you must have in mind while using iPython notebooks:

- Remember you can see, in the left side of a code cell, when was the last time it ran, if there is a number inside the keys.
- When starting a new session in the notebook, please make sure to run all cells up to the point where you last left it. Even if the output can still be viewed from the moment you ran your cells in the previews session, the kernel starts in a new state, so you will need to reload all data, etc. in a new session.
- The previous point is useful to have in mind if your answers do not match what is expected from the quizzes in the classroom. Try reloading the data and running all processing steps, one by one, to make sure you're working with the same variables and data from each step of the quizz.

## Session 1 - Data Gathering

### *Exercise 1.1*

Let's do it!! Now it's your turn to gather data. Please write bellow a Python code to access the link http://web.mta.info/developers/turnstile.html and download all files from June 2017. The file must be named turnstile_100617.txt, where 10/06/17 is the file's date.

Please see below a few commands that might help you:

Use the **urllib** library to open and redeem a webpage. Use the command below, where **url** is the webpage path to the following file:

```python
u = urllib.urlopen(url)
html = u.read()
```

Use the **BeautifulSoup** library to search for the link to the file you want to donwload in the page. Use the command below to create your *soup* object and search for all 'a' tags in the document:
 
 
```python
soup = BeautifulSoup(html, "html.parser")
links = soup.find_all('a')
```

A tip to only download the files from June is to check data in the name of the file. For instance, to donwload the 17/06/2017 file, please see if the link ends with *"turnstile_170610.txt"*. If you forget to do this, you will download all files from that page. In order to do this, you can use the following command:

```python
if '1706' in link.get('href'):
```

Our final tip is to use the command bellow to download the txt file:

```python
urllib.urlretrieve(link_do_arquivo, filename)
```

Please remember - you first have to load all packages and functions that will be used in your analysys.

In [1]:
from urllib.request import urlopen, urlretrieve
from bs4 import BeautifulSoup
from datetime import datetime as dt
import urllib.request
#your code here
import os
from datetime import datetime as dt

def get(url):
    u = urllib.request.urlopen(url)
    html = u.read()
    code = u.getcode()

    if code >= 400:
        raise ValueError("HTTP Request error %d" % code)
        
    return html, u

def check_dir(path):
    directory = os.path.abspath(path)

    if not os.path.exists(directory):
        os.makedirs(directory)

def divider():
    print ("." * 50 + "\n")
        
def download(links, basedir):
    files = []
    check_dir(basedir + '/')

    total = len(links)
    count = 0
    print ("%d files to download."%(total))
    divider()
    
    for filename, link in links.items():
        count += 1
        path = os.path.abspath(basedir + '/%s' % filename)
        files.append(path)

        print ("Start download file %s %d/%d - %2.f%%" % (path, count, total, float(count - 1) * 100 / total))
        urllib.request.urlretrieve(link, path)
        print ("Finished download file %s %d/%d - %2.f%%" % (path, count, total, float(count) * 100 / total))
        divider()

    print ("%d files download has finished successfully." %(total))
    divider()
    
    return files
    
def get_turnstile_links(year, month=None):
    try:
        html, u = get("http://web.mta.info/developers/turnstile.html")
        soup = BeautifulSoup(html, "html.parser")
        links = soup.find("div", class_="last").find_all("a")

        datalinks = {}

        for link in links:
            try:
                date = dt.strptime(link.get_text(), "%A, %B %d, %Y")

                if (date.year == year and (not month or date.month == month)):
                    href = link.get("href")
                    filename = href.rsplit("/", 1)[-1]
                    datalinks[filename] = "http://web.mta.info/developers/" + href
            except ValueError as e:
                print ("Invalid link %s: Reason %s" % (link.get("href"), str(e)))

        return datalinks
    except Exception as e:
        raise RuntimeError("Reason %s" % str(e))

input_datadir = "data"
year = 2017
month = 6
links = get_turnstile_links(year, month)
datapath = download(links, input_datadir)

4 files to download.
..................................................

Start download file C:\Users\User\Desktop\data\turnstile_170624.txt 1/4 -  0%
Finished download file C:\Users\User\Desktop\data\turnstile_170624.txt 1/4 - 25%
..................................................

Start download file C:\Users\User\Desktop\data\turnstile_170617.txt 2/4 - 25%
Finished download file C:\Users\User\Desktop\data\turnstile_170617.txt 2/4 - 50%
..................................................

Start download file C:\Users\User\Desktop\data\turnstile_170610.txt 3/4 - 50%
Finished download file C:\Users\User\Desktop\data\turnstile_170610.txt 3/4 - 75%
..................................................

Start download file C:\Users\User\Desktop\data\turnstile_170603.txt 4/4 - 75%
Finished download file C:\Users\User\Desktop\data\turnstile_170603.txt 4/4 - 100%
..................................................

4 files download has finished successfully.
......................................

### *Exercise 1.2*

Write down a function that takes the list of all names of the files you downloaded in Exercise 1.1 and compile them into one single file. There must be only one header line in the output file. 

For example, if file_1 has:
line 1...
line 2...

and the other file, file_2, has:
line 3...
line 4...
line 5...

We must combine file_1 and file_2 into one master file, as follows:

'C/A, UNIT, SCP, DATEn, TIMEn, DESCn, ENTRIESn, EXITSn'
line 1...
line 2...
line 3...
line 4...
line 5...


In [2]:
def create_master_turnstile_file(filenames, output_file):
    check_dir(output_file.rsplit("/", 1)[0])

    with open(output_file, 'w') as master_file:
        master_file.write('C/A,UNIT,SCP,STATION, LINENAME, DIVISION, DATEn,TIMEn,DESCn,ENTRIESn,EXITSn\n')
        for filename in filenames:
            with open(filename) as partial_file:
                for line in partial_file:
                    master_file.write(line)

print("Merging files...")
output_dir = "output"
turnstile_file = create_master_turnstile_file(datapath, output_dir + "/turnstile_1706.txt")
print("Files merged...")

Merging files...
Files merged...


### *Exercise 1.3*

For this exercise, you will write a function that reads the master_file created in the previous exercise and load it into a Pandas Dataframe. This function can be filtered, so that the Dataframe only has lines where column "DESCn" has the value "Regular".

For example, if the Pandas Dataframe looks like this:
    
    ,C/A,UNIT,SCP,DATEn,TIMEn,DESCn,ENTRIESn,EXITSn
    0,A002,R051,02-00-00,05-01-11,00:00:00,REGULAR,3144312,1088151
    1,A002,R051,02-00-00,05-01-11,04:00:00,DOOR,3144335,1088159
    2,A002,R051,02-00-00,05-01-11,08:00:00,REGULAR,3144353,1088177
    3,A002,R051,02-00-00,05-01-11,12:00:00,DOOR,3144424,1088231

The Dataframe must look like the following, after filtering only the lines where column DESCn has the value REGULAR:

    0,A002,R051,02-00-00,05-01-11,00:00:00,REGULAR,3144312,1088151
    2,A002,R051,02-00-00,05-01-11,08:00:00,REGULAR,3144353,1088177


In [3]:
import pandas

def filter_by_regular(filename):
    turnstile_data = pandas.read_csv(filename, dtype=object)
    turnstile_data = turnstile_data[turnstile_data['DESCn'] == 'REGULAR']
    return turnstile_data

filter_by_regular('june2017.txt')
    

C/A  UNIT       SCP      STATION  LINENAME  DIVISION       DATEn  \
1      A002  R051  02-00-00        59 ST   NQR456W       BMT  05/27/2017   
2      A002  R051  02-00-00        59 ST   NQR456W       BMT  05/27/2017   
3      A002  R051  02-00-00        59 ST   NQR456W       BMT  05/27/2017   
4      A002  R051  02-00-00        59 ST   NQR456W       BMT  05/27/2017   
5      A002  R051  02-00-00        59 ST   NQR456W       BMT  05/27/2017   
6      A002  R051  02-00-00        59 ST   NQR456W       BMT  05/27/2017   
7      A002  R051  02-00-00        59 ST   NQR456W       BMT  05/28/2017   
8      A002  R051  02-00-00        59 ST   NQR456W       BMT  05/28/2017   
9      A002  R051  02-00-00        59 ST   NQR456W       BMT  05/28/2017   
10     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/28/2017   
11     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/28/2017   
12     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/28/2017   
13     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/29/2017   
14     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/29/2017   
15     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/29/2017   
16     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/29/2017   
17     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/29/2017   
18     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/29/2017   
19     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/30/2017   
20     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/30/2017   
21     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/30/2017   
22     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/30/2017   
23     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/30/2017   
24     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/30/2017   
25     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/31/2017   
26     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/31/2017   
27     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/31/2017   
28     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/31/2017   
29     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/31/2017   
30     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/31/2017   
...     ...   ...       ...          ...       ...       ...         ...   
60503  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/28/2017   
60504  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/28/2017   
60505  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/29/2017   
60506  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/29/2017   
60507  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/29/2017   
60508  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/29/2017   
60509  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/29/2017   
60510  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/29/2017   
60511  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/30/2017   
60512  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/30/2017   
60513  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/30/2017   
60514  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/30/2017   
60515  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/30/2017   
60516  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/30/2017   
60517  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/31/2017   
60518  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/31/2017   
60519  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/31/2017   
60520  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/31/2017   
60521  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/31/2017   
60522  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/31/2017   
60523  N114  R297  01-00-03  FRANKLIN AV       ACS      

### *Exercise 1.4*


The NYC Subway data has cumulative entry and exit data in each line. Let's assume you have a Dataframe called df, which contains only lines for one particular turnstile (unique SCP, C/A, and UNIT). The following function must change these cumulative entries for counting all entries since the last reading (entries from the last line of the Dataframe).

More specifically, there are two things you should do:

1 - Create a new column, called ENTRIESn_hourly
2 - Insert in this column the difference between ENTRIESn in the current and the previous column. If the line has any NAN, fill it out/replace by 1.

Tip: The funtions shift() and fillna() in Pandas might be usefull for this exercise.

Below you will find and example of how your Dataframe should look by the end of this exercise:

        C/A  UNIT       SCP     DATEn     TIMEn    DESCn  ENTRIESn    EXITSn  ENTRIESn_hourly
    0     A002  R051  02-00-00  05-01-11  00:00:00  REGULAR   3144312   1088151                1
    1     A002  R051  02-00-00  05-01-11  04:00:00  REGULAR   3144335   1088159               23
    2     A002  R051  02-00-00  05-01-11  08:00:00  REGULAR   3144353   1088177               18
    3     A002  R051  02-00-00  05-01-11  12:00:00  REGULAR   3144424   1088231               71
    4     A002  R051  02-00-00  05-01-11  16:00:00  REGULAR   3144594   1088275              170
    5     A002  R051  02-00-00  05-01-11  20:00:00  REGULAR   3144808   1088317              214
    6     A002  R051  02-00-00  05-02-11  00:00:00  REGULAR   3144895   1088328               87
    7     A002  R051  02-00-00  05-02-11  04:00:00  REGULAR   3144905   1088331               10
    8     A002  R051  02-00-00  05-02-11  08:00:00  REGULAR   3144941   1088420               36
    9     A002  R051  02-00-00  05-02-11  12:00:00  REGULAR   3145094   1088753              153
    10    A002  R051  02-00-00  05-02-11  16:00:00  REGULAR   3145337   1088823              243

In [104]:
import pandas

def get_hourly_entries(df):
 
    diff = pandas.to_numeric(df['ENTRIESn']) - pandas.to_numeric(df['ENTRIESn'].shift(1))
    df['ENTRIESn_hourly'] = diff.fillna(1).astype(int)
        
    return df

df = filter_by_regular('june2017.txt')
get_hourly_entries(df)
    
    

C/A  UNIT       SCP      STATION  LINENAME  DIVISION       DATEn  \
1      A002  R051  02-00-00        59 ST   NQR456W       BMT  05/27/2017   
2      A002  R051  02-00-00        59 ST   NQR456W       BMT  05/27/2017   
3      A002  R051  02-00-00        59 ST   NQR456W       BMT  05/27/2017   
4      A002  R051  02-00-00        59 ST   NQR456W       BMT  05/27/2017   
5      A002  R051  02-00-00        59 ST   NQR456W       BMT  05/27/2017   
6      A002  R051  02-00-00        59 ST   NQR456W       BMT  05/27/2017   
7      A002  R051  02-00-00        59 ST   NQR456W       BMT  05/28/2017   
8      A002  R051  02-00-00        59 ST   NQR456W       BMT  05/28/2017   
9      A002  R051  02-00-00        59 ST   NQR456W       BMT  05/28/2017   
10     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/28/2017   
11     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/28/2017   
12     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/28/2017   
13     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/29/2017   
14     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/29/2017   
15     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/29/2017   
16     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/29/2017   
17     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/29/2017   
18     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/29/2017   
19     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/30/2017   
20     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/30/2017   
21     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/30/2017   
22     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/30/2017   
23     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/30/2017   
24     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/30/2017   
25     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/31/2017   
26     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/31/2017   
27     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/31/2017   
28     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/31/2017   
29     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/31/2017   
30     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/31/2017   
...     ...   ...       ...          ...       ...       ...         ...   
60503  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/28/2017   
60504  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/28/2017   
60505  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/29/2017   
60506  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/29/2017   
60507  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/29/2017   
60508  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/29/2017   
60509  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/29/2017   
60510  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/29/2017   
60511  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/30/2017   
60512  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/30/2017   
60513  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/30/2017   
60514  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/30/2017   
60515  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/30/2017   
60516  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/30/2017   
60517  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/31/2017   
60518  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/31/2017   
60519  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/31/2017   
60520  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/31/2017   
60521  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/31/2017   
60522  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/31/2017   
60523  N114  R297  01-00-03  FRANKLIN AV       ACS      

### *Exercise 1.5*

Do the same thing you did in the previous exercise, but taking into account the exits, column EXITSn.
For this, you need to create a column called EXITSn_hourly and insert the difference between the column EXITSn in the current line vs he previous line. If there is any NaN, fill it out/replace by 0.



In [22]:
import pandas

def get_hourly_exits(df):
    diff = pandas.to_numeric(df['EXITSn']) - pandas.to_numeric(df['EXITSn'].shift(1))
    df['EXITSn_hourly'] = diff.fillna(0).astype(int)
        
    return df

get_hourly_exits(df)
    

C/A  UNIT       SCP      STATION  LINENAME  DIVISION       DATEn  \
1      A002  R051  02-00-00        59 ST   NQR456W       BMT  05/27/2017   
2      A002  R051  02-00-00        59 ST   NQR456W       BMT  05/27/2017   
3      A002  R051  02-00-00        59 ST   NQR456W       BMT  05/27/2017   
4      A002  R051  02-00-00        59 ST   NQR456W       BMT  05/27/2017   
5      A002  R051  02-00-00        59 ST   NQR456W       BMT  05/27/2017   
6      A002  R051  02-00-00        59 ST   NQR456W       BMT  05/27/2017   
7      A002  R051  02-00-00        59 ST   NQR456W       BMT  05/28/2017   
8      A002  R051  02-00-00        59 ST   NQR456W       BMT  05/28/2017   
9      A002  R051  02-00-00        59 ST   NQR456W       BMT  05/28/2017   
10     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/28/2017   
11     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/28/2017   
12     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/28/2017   
13     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/29/2017   
14     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/29/2017   
15     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/29/2017   
16     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/29/2017   
17     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/29/2017   
18     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/29/2017   
19     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/30/2017   
20     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/30/2017   
21     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/30/2017   
22     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/30/2017   
23     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/30/2017   
24     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/30/2017   
25     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/31/2017   
26     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/31/2017   
27     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/31/2017   
28     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/31/2017   
29     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/31/2017   
30     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/31/2017   
...     ...   ...       ...          ...       ...       ...         ...   
60503  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/28/2017   
60504  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/28/2017   
60505  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/29/2017   
60506  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/29/2017   
60507  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/29/2017   
60508  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/29/2017   
60509  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/29/2017   
60510  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/29/2017   
60511  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/30/2017   
60512  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/30/2017   
60513  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/30/2017   
60514  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/30/2017   
60515  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/30/2017   
60516  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/30/2017   
60517  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/31/2017   
60518  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/31/2017   
60519  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/31/2017   
60520  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/31/2017   
60521  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/31/2017   
60522  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/31/2017   
60523  N114  R297  01-00-03  FRANKLIN AV       ACS      

### *Exercise 1.6*

Given an entry variable that represents time, in the format:
     "00:00:00" (hour: minutes: seconds)
    
Write a function to extract the hour part from the time in the entry variable
And return it as an integer. For example:
         
         1) if hour is 00, your code must return 0
         2) if hour is 01, your code must return 1
         3) if hour is 21, your code must return 21
        
Please return te hour as an integer.


In [23]:
def time_to_hour(time):
    
    df['hour'] = df['TIMEn'].map(lambda x: int(x.split(':')[0]))
    return df

time_to_hour(df)

C/A  UNIT       SCP      STATION  LINENAME  DIVISION       DATEn  \
1      A002  R051  02-00-00        59 ST   NQR456W       BMT  05/27/2017   
2      A002  R051  02-00-00        59 ST   NQR456W       BMT  05/27/2017   
3      A002  R051  02-00-00        59 ST   NQR456W       BMT  05/27/2017   
4      A002  R051  02-00-00        59 ST   NQR456W       BMT  05/27/2017   
5      A002  R051  02-00-00        59 ST   NQR456W       BMT  05/27/2017   
6      A002  R051  02-00-00        59 ST   NQR456W       BMT  05/27/2017   
7      A002  R051  02-00-00        59 ST   NQR456W       BMT  05/28/2017   
8      A002  R051  02-00-00        59 ST   NQR456W       BMT  05/28/2017   
9      A002  R051  02-00-00        59 ST   NQR456W       BMT  05/28/2017   
10     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/28/2017   
11     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/28/2017   
12     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/28/2017   
13     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/29/2017   
14     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/29/2017   
15     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/29/2017   
16     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/29/2017   
17     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/29/2017   
18     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/29/2017   
19     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/30/2017   
20     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/30/2017   
21     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/30/2017   
22     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/30/2017   
23     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/30/2017   
24     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/30/2017   
25     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/31/2017   
26     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/31/2017   
27     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/31/2017   
28     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/31/2017   
29     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/31/2017   
30     A002  R051  02-00-00        59 ST   NQR456W       BMT  05/31/2017   
...     ...   ...       ...          ...       ...       ...         ...   
60503  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/28/2017   
60504  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/28/2017   
60505  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/29/2017   
60506  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/29/2017   
60507  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/29/2017   
60508  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/29/2017   
60509  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/29/2017   
60510  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/29/2017   
60511  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/30/2017   
60512  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/30/2017   
60513  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/30/2017   
60514  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/30/2017   
60515  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/30/2017   
60516  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/30/2017   
60517  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/31/2017   
60518  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/31/2017   
60519  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/31/2017   
60520  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/31/2017   
60521  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/31/2017   
60522  N114  R297  01-00-03  FRANKLIN AV       ACS       IND  05/31/2017   
60523  N114  R297  01-00-03  FRANKLIN AV       ACS      

## Exercise 2 - Data Analysis

### *Exercise 2.1*

To understand the relationship between the Subway activity and the weather, please complete the data from the file already downloaded with the weather data.
We provided you with the file containing NYC weather data and made it available with the Support Material. You can access it through the link: https://s3.amazonaws.com/content.udacity-data.com/courses/ud359/turnstile_data_master_with_weather.csv

Now that we have our data in a csv file, write Python code that reads this file and saves it into a Pandas Dataframe. 

Tip: 

Use the command below to read the file:

```python
pd.read_csv('output_list.txt', sep=",")
```



In [114]:
import pandas as pd

filename = "turnstile_data_master_with_weather (3).csv"

df = pandas.read_csv(filename, dtype=object)
df.head()

Unnamed: 0  UNIT       DATEn     TIMEn Hour    DESCn ENTRIESn_hourly  \
0          0  R001  2011-05-01  01:00:00    1  REGULAR             0.0   
1          1  R001  2011-05-01  05:00:00    5  REGULAR           217.0   
2          2  R001  2011-05-01  09:00:00    9  REGULAR           890.0   
3          3  R001  2011-05-01  13:00:00   13  REGULAR          2451.0   
4          4  R001  2011-05-01  17:00:00   17  REGULAR          4400.0   

  EXITSn_hourly maxpressurei maxdewpti   ...   meandewpti meanpressurei  fog  \
0           0.0        30.31      42.0   ...         39.0         30.27  0.0   
1         553.0        30.31      42.0   ...         39.0         30.27  0.0   
2        1262.0        30.31      42.0   ...         39.0         30.27  0.0   
3        3708.0        30.31      42.0   ...         39.0         30.27  0.0   
4        2501.0        30.31      42.0   ...         39.0         30.27  0.0   

  rain meanwindspdi mintempi meantempi maxtempi precipi thunder  
0  0.0          5.0     50.0      60.0     69.0     0.0     0.0  
1  0.0          5.0     50.0      60.0     69.0     0.0     0.0  
2  0.0          5.0     50.0      60.0     69.0     0.0     0.0  
3  0.0          5.0     50.0      60.0     69.0     0.0     0.0  
4  0.0          5.0     50.0      60.0     69.0     0.0     0.0  

[5 rows x 22 columns]

### *Exercise 2.2*

Now, create a function that calculates the number of rainy days. For this, return the count of the number of days where the column *"rain"* is equal to 1.

Tip: You might think that interpreting numbers as integers or floats might not
     work at first. To handle this issue, it might be useful to convert
     these numbers into integers. You can do this by writting cast (column as integer).
     So, for example, if we want to launch the column maxtempi as an integer, we have to
     write something like cast (maxtempi as integer) = 76, instead of just
     where maxtempi = 76.

In [94]:


try:
    import pymysql
    pymysql.install_as_MySQLdb()
except ImportError:
    pass
import pandasql as pdsql

def num_rainy_days(df):
    df['rain'] = pandas.to_numeric(df['rain'])
    num = len(df[df.rain == 1].groupby('DATEn'))
        
    return num

num_rainy_days(df)
    


    
    

10

### *Exercise 2.3*

Calculate if the day was cloudy or not (0 or 1) and the maximum temperature for fog (i.e. the maximum temperature 
     for cloudy days).

In [115]:
def max_temp_aggregate_by_fog(df):
    query = "SELECT fog, MAX(maxtempi) FROM df GROUP BY fog"
    return pdsql.sqldf(query.lower(), locals())

max_temp_aggregate_by_fog(df)

fog max(maxtempi)
0  0.0          86.0
1  1.0          81.0

### *Exercise 2.4

Now, calculate the mean for 'meantempi' for the days that are Saturdays or Sundays (weekend):

In [116]:
def avg_weekend_temperature(filename):
    df = pandas.read_csv(filename, dtype=object)
    df['weekday'] = pd.to_datetime(df['DATEn']).dt.dayofweek
    filtered_data = df['meantempi'].astype(float)[df['weekday'] >= 5]
    mean_temp_weekends = filtered_data.mean()
    return mean_temp_weekends

avg_weekend_temperature("turnstile_data_master_with_weather (3).csv")

65.10066685403307

### *Exercise 2.5

Calculate the mean of the minimum temperature 'mintempi' for the days when the minimum temperature was greater that 55 degrees:

In [118]:
def avg_min_temperature(filename):

    df = pandas.read_csv(filename, dtype=object)
    filtered_data = df['mintempi'].astype(float)[df['rain'] == '1.0'][df['mintempi'].astype(float) > 55]
    avg_mintemp_on_rainy_days = filtered_data.mean()
    
    avg_min_temp_rainy = avg_mintemp_on_rainy_days
    
    return avg_min_temp_rainy

avg_min_temperature("turnstile_data_master_with_weather (3).csv")

61.238860398860396

### *Exercise 2.6

Before you make any analysis, it might be useful to look at the data we want to analyse. More specifically, we will evaluate the entries by hour in our data from the NYC Subway to determine the data distribution. This data is stored in the column ['ENTRIESn_hourly'].
    
Draw two histogramns in the same axis, to show the entries when it's raining vs when it's not. 
Below, you will find an example of how to draw histogramns with Pandas and Matplotlib:
     
```python
Turnstile_weather ['column_to_graph']. Hist ()
```   
    

In [ ]:
from datetime import datetime as dt
from datetime import date
import numpy as np
import pandas
import matplotlib.pyplot as plt

%matplotlib inline  

def entries_histogram(turnstile_weather):

    plt.figure() #create a plot object which we will popluate with histograms
    turnstile_weather['ENTRIESn_hourly'] = pandas.to_numeric(turnstile_weather['ENTRIESn_hourly'])
    turnstile_weather['rain'] = pandas.to_numeric(turnstile_weather['rain'])
    
    turnstile_weather[turnstile_weather['rain'] == 0]['ENTRIESn_hourly'].hist(alpha = 0.4)
    turnstile_weather['ENTRIESn_hourly'][turnstile_weather.rain == 1].hist(alpha = 0.4)
    plt.legend(('rainining','not raining'))
    plt.xlabel("Entries per hour")
    plt.ylabel("FREQUENCY")
    return plt

filename = "turnstile_data_master_with_weather (3).csv"
df = pandas.DataFrame(pandas.read_csv(filename, dtype=object))
entries_histogram(df)

### *Exercise 2.7

The data you just plotted is in what kind of distribution? Is there a difference in distribution between rainy and non-rainy days?

** Answer **: The plot here obtained in 2.6 is skewed distribution. Its a positive skewed distribution. Yes, there is a difference in rainy and non-rainy days distribution. The entries in rainy day is more than that of non-rainy days.

### *Exercise 2.8

Build a function that returns:

1. The mean of entries when it's raining
2. The mean of entries when it's not raining




In [122]:
import numpy as np

import pandas

def means(turnstile_weather):
    
    
    ### YOUR CODE HERE ###
    query = """
        SELECT rain, avg(ENTRIESn_hourly) AS avg_entries_hourly FROM turnstile_weather group by rain
    """
    
    sqldf = pdsql.sqldf(query.lower(), locals())
    with_rain_mean = sqldf['avg_entries_hourly'][sqldf['rain'] == 1].iloc[0]
    without_rain_mean = sqldf['avg_entries_hourly'][sqldf['rain'] == 0].loc[0]
    p = None
    
    return with_rain_mean, without_rain_mean, p # leave this line for the grader
    
means(df)

(1105.4463767458733, 1090.278780151855, None)

Answer to the following questions according to your functions' exits:

1. What is the mean of entries when it's raining?
2. What is the mean of entries when it's not raining?


** Answer **:1.The average of the entries in rainy days is 1105.4463 which after rounding of turns to be 1105.45
             2.The average of the entries in non rainy days is 1090.2787 after rounding of it turns to be 1090.28

## Exercise 3 - Map Reduce

### *Exercise 3.1*

The entry for this exercise is the same file from the previous session (Exercise 2). You can download the file from this link:

 https://s3.amazonaws.com/content.udacity-data.com/courses/ud359/turnstile_data_master_with_weather.csv

Now, we will create a mapper. For each entry line, the mapper exit must PRINT (not return) UNIT as a key, and the number of ENTRIESn_hourly as the value. Separate the key and the value with a tab. For example: 'R002 \ t105105.0'

Export your mapper into a file named mapper_result.txt and send it with your submission. The code for exporting your mapper is already written in the code bellow.




In [47]:
import sys

def mapper():
    for line in sys.stdin:
        data = line.strip().split(",")
        
        if len(data) >= 7:
            unit = data[1]
            entries = data[6]
        
            print(f"{unit}\t{entries}")

sys.stdin = open('turnstile_data_master_with_weather.csv')
sys.stdout = open('mapper_result.txt', 'w')
mapper()
    

    

### *Exercise 3.2*

Now, create the reducer. Given the mapper result from the previous exercise, the reducer must print (not return) one line per UNIT, with the total number of ENTRIESn_hourly during May (which is our data duration), separated by a tab. An example of exit line from the reducer may look like this: 'R001 \ t500625.0'

You can assume that the entry for the reducer is ordered in a way that all lines corresponding to a particular unit are grouped. However, the reducer exit will have repetition, as there are stores that appear in different files' locations.

Export your reducer into a file named reducer_result.txt and send it with your submission.

In [60]:
def reducer():
    

    for line in sys.stdin:
        entriesTotal = 0
    oldKey = None

    for line in sys.stdin:
        data_mapped = line.strip().split("\t")
        if len(data_mapped) != 2:
            continue
        
        thisKey, thisEntry = data_mapped
        
        try:
            thisEntry_to_f = float(thisEntry)
            if oldKey and oldKey != thisKey:
                print(f"{oldKey}\t{entriesTotal}")
                oldKey = thisKey
                entriesTotal = 0
            
            oldKey = thisKey
            entriesTotal += thisEntry_to_f
        except ValueError:
            pass
            
sys.stdin = open('mapper_result.txt')
sys.stdout = open('reducer_result.txt', 'w')
reducer()
        